Cloud private code host support docs (#49748)
sourcegraph/sourcegraph: doc/cloud/index.md

More details about the locations and data storage can be found in [our handbook](https://handbook.sourcegraph.com/departments/cloud/technical-docs/multi-region/)

### Private Code Host support

> Sourcegraph Cloud connects to customer code hosts from 2 public NAT IPs. Customers can add Sourcegraph customer-dedicated IPs to their IP allowlist.

Reviewer: ```suggestion
Sourcegraph Cloud connects to customer code hosts from 2 public NAT IPs. Customers can add the dedicated IPs for their Sourcegraph Cloud instance to an IP allowlist on their private code host.
```

#### Code host on AWS without public access

> As part of enterprise tier, for customers having code host without public access deployed on AWS, Sourcegraph Cloud offers [high available site to site VPN solution](https://cloud.google.com/network-connectivity/docs/vpn/tutorials/create-ha-vpn-connections-google-cloud-aws) with [AWS Private Link](https://docs.aws.amazon.com/vpc/latest/privatelink/what-is-privatelink.html) inside AWS network.

Reviewer: ```suggestion
As part of the [Enterprise tier](https://about.sourcegraph.com/pricing), Sourcegraph Cloud offers customers that have code hosts without public access deployed on AWS a [highly available site-to-site VPN solution](https://cloud.google.com/network-connectivity/docs/vpn/tutorials/create-ha-vpn-connections-google-cloud-aws) with [AWS Private Link](https://docs.aws.amazon.com/vpc/latest/privatelink/what-is-privatelink.html) inside AWS's network, so that access to a private code host never occurs over the public internet.
```

> Solution architecture

Reviewer: ```suggestion
Solution architecture:
```

<img src="https://sourcegraphstatic.com/private-code-host-solution-vpn-aws-private-link.png" class="screenshot">

> Site to site GCP with AWS VPN advantages:

Reviewer: ```suggestion
Advantages of the site-to-site GCP to AWS VPN include:

```

- encrypted connection between Sourcegraph Cloud and customer code host
- multiple tunnels to provide high availability between Cloud
instance and customer code host

> AWS Private Link advantages:

Reviewer: ```suggestion
Advantages of AWS Private Link include:

```

- ensures customer only opens connectivity to their code host inside AWS network
- provides customer mechanism to select AWS Princincal (AWS Account or more granular), who can connect to their code host
- allows customer to control incoming connections
- supports private DNS

> To learn more about Private Code Host details and how customers can expose their private code host via AWS Private Link, refer to [our handbook](https://handbook.sourcegraph.com/departments/cloud/technical-docs/private-code-hosts).

Reviewer: > how customers can expose their private code host via AWS Private Link, refer to [our handbook]

IMO we shouldn't link handbook pages as official support docs, and rather more "here's how we operate X internally, check it out if you are interested" - maybe we should move the customer setup guide into docs.sourcegraph.com

Reviewer: Also customers navigating to Cloud pages in the handbook will see an automatic banner that warns that handbook content is not to be used as source-of-truth user-facing documentation, because it has a lot of internal references and we play pretty loose with handbook docs:

![image](https://user-images.githubusercontent.com/23356519/226564462-bffadffd-f80f-4516-92a8-7d0364cd0969.png)

Author: Agree 👍 updated docs not to link handbook guide from here.

More details about the locations and data storage can be found in [our handbook](https://handbook.sourcegraph.com/departments/cloud/technical-docs/multi-region/)

> ### Private Code Host support

Reviewer 2: ```suggestion
### Private Code Host support (AWS)
```

Sourcegraph Cloud connects to customer code hosts from 2 public NAT IPs. Customers can add the dedicated IPs for their Sourcegraph Cloud instance to an IP allowlist on their private code host.

#### Code host on AWS without public access

As part of the [Enterprise tier](https://about.sourcegraph.com/pricing), Sourcegraph Cloud offers customers that have code hosts without public access deployed on AWS a [highly available site-to-site VPN solution](https://cloud.google.com/network-connectivity/docs/vpn/tutorials/create-ha-vpn-connections-google-cloud-aws) with [AWS Private Link](https://docs.aws.amazon.com/vpc/latest/privatelink/what-is-privatelink.html) inside AWS's network, so that access to a private code host never occurs over the public internet.

Solution architecture:
<img src="https://sourcegraphstatic.com/private-code-host-solution-vpn-aws-private-link.png" class="screenshot">

Advantages of the site-to-site GCP to AWS VPN include:
- encrypted connection between Sourcegraph Cloud and customer code host
- multiple tunnels to provide high availability between Cloud
instance and customer code host

Advantages of AWS Private Link include:

> - ensures customer only opens connectivity to their code host inside AWS network

Reviewer 2: ```suggestion
- connectivity to customer VPC is only available inside AWS network
```

> - provides customer mechanism to select AWS Princincal (AWS Account or more granular), who can connect to their code host

Reviewer 2: ```suggestion
-  ability to select AWS Principal (AWS Account or more granular) that can connect to customer code host
```
